flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/bundle/AbstractMapBundleOperator.java

```diff
@@ -0,0 +1,210 @@
+
+	private static final long serialVersionUID = 5081841938324118594L;
+
+	/** The map in heap to store elements. */
+	private final transient Map<K, V> bundle;
+
+	/** The trigger that determines how many elements should be put into a bundle. */
+	private final BundleTrigger<IN> bundleTrigger;
+
+	/** The function used to process when receiving element. */
+	private final MapBundleFunction<K, V, IN, OUT> function;
+
+	/** Output for stream records. */
+	private transient Collector<OUT> collector;
+
+	private transient int numOfElements = 0;
+
+	private transient volatile boolean isInFinishingBundle = false;
+	private transient Object checkpointingLock;
+
+	protected AbstractMapBundleOperator(
+			MapBundleFunction<K, V, IN, OUT> function,
+			BundleTrigger<IN> bundleTrigger) {
+		chainingStrategy = ChainingStrategy.ALWAYS;
+		this.bundle = new HashMap<>();
+		this.function = checkNotNull(function, "function is null");
+		this.bundleTrigger = checkNotNull(bundleTrigger, "bundleTrigger is null");
+	}
+
+	@Override
+	public void setup(
+			StreamTask<?, ?> containingTask,
+			StreamConfig config,
+			Output<StreamRecord<OUT>> output) {
+		super.setup(containingTask, config, output);
+		this.checkpointingLock = getContainingTask().getCheckpointLock();
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		function.open(new ExecutionContextImpl(this, getRuntimeContext()));
+
+		this.numOfElements = 0;
+		this.collector = new StreamRecordCollector<>(output);
+
+		bundleTrigger.registerCallback(this);
+		// reset trigger
+		bundleTrigger.reset();
+		LOG.info("BundleOperator's trigger info: " + bundleTrigger.explain());
+
+		// counter metric to get the size of bundle
+		getRuntimeContext().getMetricGroup().gauge("bundleSize", (Gauge<Integer>) () -> numOfElements);
+		getRuntimeContext().getMetricGroup().gauge("bundleRatio", (Gauge<Double>) () -> {
+			int numOfKeys = bundle.size();
+			if (numOfKeys == 0) {
+				return 0.0;
+			} else {
+				return 1.0 * numOfElements / numOfKeys;
+			}
+		});
+	}
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+
+		// get the key and value for the map bundle
+		final IN input = element.getValue();
+		final K bundleKey = getKey(input);
+		final V bundleValue = bundle.get(bundleKey);
+
+		// get a new value after adding this element to bundle
+		final V newBundleValue = function.addInput(bundleValue, input);
+
+		// update to map bundle
+		bundle.put(bundleKey, newBundleValue);
+
+		numOfElements++;
+		bundleTrigger.onElement(input);
+	}
+
+	/**
+	 * Get the key for current processing element, which will be used as the map bundle's key.
+	 */
+	protected abstract K getKey(final IN input) throws Exception;
+
+	@Override
+	public void finishBundle() throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+		isInFinishingBundle = true;
```

Comment:
why need `isInFinishingBundle` to wait `checkpointingLock `? which other thread will invoke `finishBundle`?
Should Flink guarantee thread security for `processElement`, `processWatermark`, `prepareSnapshotPreBarrier`?

Reply:
Yes, i will add more comments to explain why we need this.

Reply:
We don't need this as long as we only have count bundle trigger. After we introduce time based trigger, this will be needed. I will delete them for now.